src/lib/midiemu/MidiAudioSourceStream.pas

```diff
@@ -0,0 +1,238 @@
+    function ReadData(Buffer: PByte; BufferSize: integer): integer; override;
+    function GetAudioFormatInfo: TAudioFormatInfo; override;
+    procedure Close; override;
+    procedure HandleMidiEvent(MidiMessage: byte; Data1: byte; Data2: byte);
+  end;
+
+implementation
+
+constructor TMidiAudioSourceStream.Create(MidiFile: Pointer; Format: TAudioFormatInfo);
+begin
+  FFormat := Format.Copy;
+  FEnvLevel := 0;
+  FPhase := 0;
+  FActive := False;
+  FRelease := False;
+  FPosition := 0;
+  FNote := -1;
+  FFreq := 440.0;
+  FVel := 0.0;
+  FOvertoneLevel := 0.0;
+  FOvertonePhase := 0.0;
+  FLock := TCriticalSection.Create;
+  Log.LogDebug('MidiAudioSourceStream: Created', 'MidiSynth');
+end;
+
+function TMidiAudioSourceStream.IsEOF: boolean;
+begin
+  Result := False;
+end;
+
+function TMidiAudioSourceStream.IsError: boolean;
+begin
+  Result := false;
+end;
+
+function TMidiAudioSourceStream.GetLength: real;
+begin
+  Result := 0; // Not implemented
+end;
+
+function TMidiAudioSourceStream.GetPosition: real;
+begin
+  Result := FPosition;
+end;
+
+procedure TMidiAudioSourceStream.SetPosition(Time: real);
+begin
+  FPosition := Time;
+end;
+
+function TMidiAudioSourceStream.ReadData(Buffer: PByte; BufferSize: integer): integer;
+
+var
+  SampleRate: Integer;
+  Channels: Integer;
+  Frames: Integer;
+  s: PSmallInt;
+  i, j: Integer;
+  phaseInc, overtoneInc, sample, overtone: Double;
+  attackStep, releaseStep: Double;
+  outSample: SmallInt;
+  tempSample: Int64;
+begin
+  SampleRate := Trunc(FFormat.SampleRate);
+  Channels := Trunc(FFormat.Channels);
+  Frames := BufferSize div (Channels * 2); // 2 bytes per sample
+  s := PSmallInt(Buffer);
+  attackStep := 1.0 / (0.005 * SampleRate);  // 5ms attack
+  releaseStep := 1.0 / (0.005 * SampleRate); // 5ms release
+  phaseInc := 2 * Pi * FFreq / SampleRate;
+  overtoneInc := 4 * Pi * FFreq / SampleRate; // 2nd harmonic
+
+  FLock.Enter;
+  try
+    for i := 0 to Frames - 1 do
+    begin
+      // Envelope
+      if FRelease then
+      begin
+        FEnvLevel := FEnvLevel - releaseStep;
+        if FEnvLevel <= 0 then
+        begin
+          FEnvLevel := 0;
+          FActive := False;
+        end;
+      end
+      else if FEnvLevel < 1.0 then
+      begin
+        FEnvLevel := FEnvLevel + attackStep;
+        if FEnvLevel > 1.0 then FEnvLevel := 1.0;
+      end;
+
+      if FActive then
+      begin
+        sample := Sin(FPhase) * FVel * FEnvLevel;
+        // Add overtone with exponential decay for note restart detection
+        overtone := Sin(FOvertonePhase) * FOvertoneLevel * FEnvLevel;
+        sample := sample + overtone;
+        // Decay overtone
+        FOvertoneLevel := FOvertoneLevel * 0.995;
+        // FPhase handling: keep between -Pi and +Pi
+        FPhase := FPhase + phaseInc;
+        if FPhase > Pi then FPhase := FPhase - 2 * Pi;
+        if FPhase < -Pi then FPhase := FPhase + 2 * Pi;
```

Comment:
FPhase will never be < -Pi because phaseInc is always positive and when we subtract from FPhase in line 151, it ends up being > -Pi.